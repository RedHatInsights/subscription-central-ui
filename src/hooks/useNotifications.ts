import React, { useContext } from 'react';
import {
  NotificationContext,
  NotificationVariantType,
  NotificationOptions
} from '../contexts/NotificationProvider';

interface Notification {
  variant: NotificationVariantType;
  message: string;
  key: string;
  timeout?: boolean;
  actionLinks?: React.ReactNode;
  downloadHref?: string;
}

interface Notifications {
  notifications: Notification[];
  addSuccessNotification(message: string, options?: NotificationOptions): string;
  addErrorNotification(message: string, options?: NotificationOptions): string;
  addInfoNotification(message: string, options?: NotificationOptions): string;
  removeNotification(key: string, options?: NotificationOptions): void;
}

const useNotifications = (): Notifications => {
  const { notifications, addNotification, removeNotification } = useContext(NotificationContext);

  const addSuccessNotification = (message: string, options?: NotificationOptions): string => {
    return addNotification('success', message, options);
  };

  const addErrorNotification = (message: string, options?: NotificationOptions): string => {
    return addNotification('danger', message, options);
  };

  const addInfoNotification = (message: string, options?: NotificationOptions): string => {
    return addNotification('info', message, options);
  };

  return {
    notifications,
    addSuccessNotification,
    addErrorNotification,
    addInfoNotification,
    removeNotification
  };
};

export default useNotifications;
