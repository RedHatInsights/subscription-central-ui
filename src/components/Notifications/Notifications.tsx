import React, { FC } from 'react';
import { Alert, AlertActionCloseButton, AlertGroup } from '@patternfly/react-core';
import useNotifications from '../../hooks/useNotifications';

const Notifications: FC = () => {
  const { notifications, removeNotification } = useNotifications();

  return (
    <AlertGroup isToast>
      {notifications.map((notification) => (
        <Alert
          isLiveRegion
          timeout={true}
          title={notification.message}
          variant={notification.variant}
          key={notification.key}
          actionClose={
            <AlertActionCloseButton
              title={notification.message}
              variantLabel={`${notification.variant} alert`}
              onClose={() => removeNotification(notification.key)}
            />
          }
        />
      ))}
    </AlertGroup>
  );
};

export default Notifications;
