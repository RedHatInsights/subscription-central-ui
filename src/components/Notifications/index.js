import Notifications from './Notifications';

export { Notifications as default };
