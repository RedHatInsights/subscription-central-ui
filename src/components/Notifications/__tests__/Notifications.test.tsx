import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import Notifications from '../Notifications';
import useNotifications from '../../../hooks/useNotifications';

jest.mock('../../../hooks/useNotifications');

describe('Notifications', () => {
  it('renders correctly', () => {
    (useNotifications as jest.Mock).mockReturnValue({
      notifications: [
        { message: 'Some stuff went well!', variant: 'success', key: '123' },
        { message: 'Some other stuff did not.', variant: 'danger', key: '456' },
        { message: 'And this happened.', variant: 'info', key: '789' }
      ]
    });
    render(<Notifications />);
    expect(document.body).toMatchSnapshot();
  });

  it('calls the remove notifications method on click of the X', () => {
    const mockRemoveNotification = jest.fn();
    (useNotifications as jest.Mock).mockReturnValue({
      notifications: [{ message: 'Some stuff went well!', variant: 'success', key: '123' }],
      removeNotification: mockRemoveNotification
    });
    const { getByTestId } = render(<Notifications />);
    fireEvent.click(getByTestId('notification-close-btn-0'));
    expect(mockRemoveNotification).toHaveBeenCalledTimes(1);
  });

  it('revokes object URL if downloadHref is passed across and user clicks X', () => {
    const mockRemoveNotification = jest.fn();
    window.URL.revokeObjectURL = jest.fn();

    (useNotifications as jest.Mock).mockReturnValue({
      notifications: [
        {
          message: 'Some stuff went well!',
          variant: 'success',
          key: '123',
          downloadHref: 'foo.com'
        }
      ],
      removeNotification: mockRemoveNotification
    });
    const { getByTestId } = render(<Notifications />);
    fireEvent.click(getByTestId('notification-close-btn-0'));
    expect(window.URL.revokeObjectURL).toHaveBeenCalledTimes(1);
  });
});
