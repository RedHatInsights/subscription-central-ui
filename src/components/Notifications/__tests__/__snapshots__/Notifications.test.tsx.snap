// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notifications renders correctly 1`] = `
<body>
  <div />
  <div>
    <ul
      class="pf-c-alert-group pf-m-toast"
    >
      <li>
        <div
          aria-atomic="false"
          aria-label="Success Alert"
          aria-live="polite"
          class="pf-c-alert pf-m-success"
          data-ouia-component-id="OUIA-Generated-Alert-success-1"
          data-ouia-component-type="PF4/Alert"
          data-ouia-safe="true"
        >
          <div
            class="pf-c-alert__icon"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 512 512"
              width="1em"
            >
              <path
                d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
              />
            </svg>
          </div>
          <h4
            class="pf-c-alert__title"
          >
            <span
              class="pf-u-screen-reader"
            >
              Success alert:
            </span>
            Some stuff went well!
          </h4>
          <div
            class="pf-c-alert__action"
          >
            <button
              aria-disabled="false"
              aria-label="Close success alert alert: Some stuff went well!"
              class="pf-c-button pf-m-plain"
              data-ouia-component-id="OUIA-Generated-Button-plain-1"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe="true"
              title="Some stuff went well!"
              type="button"
            >
              <svg
                aria-hidden="true"
                fill="currentColor"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 352 512"
                width="1em"
              >
                <path
                  d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                />
              </svg>
            </button>
          </div>
          
        </div>
      </li>
      <li>
        <div
          aria-atomic="false"
          aria-label="Danger Alert"
          aria-live="polite"
          class="pf-c-alert pf-m-danger"
          data-ouia-component-id="OUIA-Generated-Alert-danger-1"
          data-ouia-component-type="PF4/Alert"
          data-ouia-safe="true"
        >
          <div
            class="pf-c-alert__icon"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 512 512"
              width="1em"
            >
              <path
                d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zm-248 50c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
              />
            </svg>
          </div>
          <h4
            class="pf-c-alert__title"
          >
            <span
              class="pf-u-screen-reader"
            >
              Danger alert:
            </span>
            Some other stuff did not.
          </h4>
          <div
            class="pf-c-alert__action"
          >
            <button
              aria-disabled="false"
              aria-label="Close danger alert alert: Some other stuff did not."
              class="pf-c-button pf-m-plain"
              data-ouia-component-id="OUIA-Generated-Button-plain-2"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe="true"
              title="Some other stuff did not."
              type="button"
            >
              <svg
                aria-hidden="true"
                fill="currentColor"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 352 512"
                width="1em"
              >
                <path
                  d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                />
              </svg>
            </button>
          </div>
          
        </div>
      </li>
      <li>
        <div
          aria-atomic="false"
          aria-label="Info Alert"
          aria-live="polite"
          class="pf-c-alert pf-m-info"
          data-ouia-component-id="OUIA-Generated-Alert-info-1"
          data-ouia-component-type="PF4/Alert"
          data-ouia-safe="true"
        >
          <div
            class="pf-c-alert__icon"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 512 512"
              width="1em"
            >
              <path
                d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 110c23.196 0 42 18.804 42 42s-18.804 42-42 42-42-18.804-42-42 18.804-42 42-42zm56 254c0 6.627-5.373 12-12 12h-88c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h12v-64h-12c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h64c6.627 0 12 5.373 12 12v100h12c6.627 0 12 5.373 12 12v24z"
              />
            </svg>
          </div>
          <h4
            class="pf-c-alert__title"
          >
            <span
              class="pf-u-screen-reader"
            >
              Info alert:
            </span>
            And this happened.
          </h4>
          <div
            class="pf-c-alert__action"
          >
            <button
              aria-disabled="false"
              aria-label="Close info alert alert: And this happened."
              class="pf-c-button pf-m-plain"
              data-ouia-component-id="OUIA-Generated-Button-plain-3"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe="true"
              title="And this happened."
              type="button"
            >
              <svg
                aria-hidden="true"
                fill="currentColor"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 352 512"
                width="1em"
              >
                <path
                  d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                />
              </svg>
            </button>
          </div>
          
        </div>
      </li>
    </ul>
  </div>
</body>
`;
